src/main/java/com/qubit/solution/fenixedu/integration/cgd/services/utils/MoreThanOneAssociatedUnitStrategy.java

```diff
@@ -0,0 +1,79 @@
+package com.qubit.solution.fenixedu.integration.cgd.services.utils;
+
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.LinkedList;
+import java.util.List;
+import java.util.Set;
+
+import org.fenixedu.academic.domain.organizationalStructure.Unit;
+import org.fenixedu.academic.domain.student.Registration;
+
+import com.qubit.solution.fenixedu.integration.cgd.domain.configuration.CgdIntegrationConfiguration;
+
+public class MoreThanOneAssociatedUnitStrategy extends BaseIESCodeProviderStrategy implements CgdIESCodeProviderStrategyClass {
+
+    @Override
+    public List<String> getIESCode(Registration registration) {
+        Unit associatedUnit = registration.getDegree().getUnit();
+
+        if (associatedUnit == null) {
+            return super.getIESCode(registration);
+        }
+
+        return CgdIntegrationConfiguration.getInstance().getAllowsMultipleUnits() ? findAllUnitsCodes(registration,
+                associatedUnit) : findFirstUnitFromConfiguration(registration, associatedUnit);
+    }
+
+    private List<String> findAllUnitsCodes(Registration registration, Unit associatedUnit) {
+        Set<Unit> allowedUnits = CgdIntegrationConfiguration.getInstance().getUnitsSet();
+        List<String> result = new ArrayList<String>();
+
+        if (allowedUnits.contains(associatedUnit)) {
+            String code = associatedUnit.getCode();
+            if (!result.contains(code)) {
+                result.add(code);
+            }
+        }
+
+        continueSearchUnitWithParentUnits(associatedUnit.getParentUnits(), allowedUnits, result);
+
+        return result.isEmpty() ? super.getIESCode(registration) : result;
+    }
+
+    private void continueSearchUnitWithParentUnits(Collection<Unit> associatedParentUnits, Set<Unit> allowedUnitsForSearch,
+            List<String> listToReturn) {
+        associatedParentUnits.forEach(unit -> {
```

Comment:
@miguel-morais remember that a forEach with an if is basically a stream with a filter, considering the previous suggestion of having the listToReturn as a Set this can be simplified to:

`
associatedParentUnits.stream().filter(unit -> allowedUnitsForSearch.contains(unit)).forEach(unit -> listToReturn.add(unit.getCode());
`

Reply:
I agree with the suggestion, however since I want to search every unit, and it's parents, applying the filter will make me skip units in the search:
Allowed4                
NotAllowed2 (Parents: Allowed4 and NotAllowed5)
Allowed1 (Parents: NotAllowed2 and Allowed3)                   
NotAllowed5
Allowed3 (Parents: Allowed6)
Allowed6
Assuming the following ""tree"". We see that applying the filter will make me skip unit Allowed4 because NotAllowed2 is not allowed.
The only way that I see to use the filter is doing in two lines:

```
associatedParentUnits.stream().filter(unit -> allowedUnitsForSearch.contains(unit)).forEach(unit -> listToReturn.add(unit.getCode());
associatedParentUnits.stream().forEach(unit -> continueSearchUnitWithParentUnits(unit.getAllParentUnits(), allowedUnitsForSearch, listToReturn));
```

Reply:
You are correct, but looking into the domain I saw a way of make things even easier which is using the method `getAllParentUnits()`. Using this, you no longer need this helper methods.